src/bootstrap/test.rs

```diff
@@ -1118,7 +1118,11 @@ impl Step for Tidy {
         cmd.arg(&builder.src);
         cmd.arg(&builder.initial_cargo);
         cmd.arg(&builder.out);
-        cmd.arg(builder.jobs().to_string());
+        // Tidy is heavily IO constrained. Still respect `-j`, but use a higher limit if `jobs` hasn't been configured.
+        let jobs = builder.config.jobs.unwrap_or_else(|| {
+            8 * std::thread::available_parallelism().map_or(1, std::num::NonZeroUsize::get) as u32
+        });
```

Comment:
This might not be the best idea since it only controls coarse high-level parallelism in `main` and some tests run faster than others. On a many-core machine that'll bump the limit into the hundreds and have no practical effect because then the longest-running test will just dominate.

Instead it might be better to do multi-threaded directory traversal which `ignore` supports. But I'm not sure how we should cap that, sadly ignore doesn't support a shared thread-pool.

Reply:
If I understand correctly, you're saying on machines with many cores this won't have much of an improvement over the existing job limit? That seems ok to me - I mostly wrote this for machines with few cores, where this has a really noticeable difference. High-powered machines are already finishing in under a second so I'm not too worried about improving their times :)

Reply:
Yeah it's not really a problem, it just feels inelegant. As an incremental change it _maybe_ makes sense. If we were building tidy from scratch it wouldn't.

Did you make incremental measurements of the factor? In the rustfmt part of the tidy command I'm only using a 2x factor, not 8x to cover up IO wait times.

https://github.com/rust-lang/rust/blob/e350fe4e608b653da47e8012d13ef701613e717b/src/bootstrap/format.rs#L256-L258

Reply:
It probably also helps that the format step is separating directory iteration (lots of IO) from file-processing (a bit IO, lots of byte-munching)

Reply:
On my 16-core (Ryzen 7700) linux machine:

```
# 128 jobs
Benchmark 1: /home/jyn/src/rust2/build/x86_64-unknown-linux-gnu/stage0-tools-bin/rust-tidy /home/jyn/src/rust2 /home/jyn/src/rust2/build/x86_64-unknown-linux-gnu/stage0/bin/cargo /home/jyn/src/rust2/build 128
  Time (mean ± σ):     237.7 ms ±  13.0 ms    [User: 828.6 ms, System: 380.4 ms]
  Range (min … max):   223.3 ms … 264.5 ms    13 runs

# 16 jobs
Benchmark 1: /home/jyn/src/rust2/build/x86_64-unknown-linux-gnu/stage0-tools-bin/rust-tidy /home/jyn/src/rust2 /home/jyn/src/rust2/build/x86_64-unknown-linux-gnu/stage0/bin/cargo /home/jyn/src/rust2/build 16
  Time (mean ± σ):     271.3 ms ±   6.1 ms    [User: 786.5 ms, System: 359.5 ms]
  Range (min … max):   259.0 ms … 277.7 ms    10 runs

# 2 jobs
Benchmark 1: /home/jyn/src/rust2/build/x86_64-unknown-linux-gnu/stage0-tools-bin/rust-tidy /home/jyn/src/rust2 /home/jyn/src/rust2/build/x86_64-unknown-linux-gnu/stage0/bin/cargo /home/jyn/src/rust2/build 2
  Time (mean ± σ):     773.4 ms ±  17.6 ms    [User: 718.7 ms, System: 327.3 ms]
  Range (min … max):   755.9 ms … 819.6 ms    10 runs
```

on my thinkpad from 2020 (8th gen i7, 8 cores, probably running windows update in the background):
```
# 64 jobs
Benchmark 1: build\\x86_64-pc-windows-msvc\\stage0-tools-bin\\rust-tidy.exe . build\\x86_64-pc-windows-msvc\\stage0/bin/cargo build 64
  Time (mean ± σ):      4.147 s ±  0.756 s    [User: 5.673 s, System: 9.183 s]
  Range (min … max):    2.769 s …  4.956 s    10 runs

# 8 jobs
Benchmark 1: build\\x86_64-pc-windows-msvc\\stage0-tools-bin\\rust-tidy.exe . build\\x86_64-pc-windows-msvc\\stage0/bin/cargo build 8
  Time (mean ± σ):      6.123 s ±  0.784 s    [User: 4.189 s, System: 7.575 s]
  Range (min … max):    4.709 s …  7.793 s    10 runs

# 2 jobs
Benchmark 1: build\\x86_64-pc-windows-msvc\\stage0-tools-bin\\rust-tidy.exe . build\\x86_64-pc-windows-msvc\\stage0/bin/cargo build 2
  Time (mean ± σ):      8.173 s ±  0.372 s    [User: 3.342 s, System: 6.745 s]
  Range (min … max):    7.619 s …  8.816 s    10 runs
```
So I think this will be an improvement on both fast and slow machines. I would rather not spend too much more time on it in this PR, we can always improve it later.

Reply:
oh, I missed that you said an incremental factor, sorry. on the fast machine (32 jobs):
```
Benchmark 1: /home/jyn/src/rust2/build/x86_64-unknown-linux-gnu/stage0-tools-bin/rust-tidy /home/jyn/src/rust2 /home/jyn/src/rust2/build/x86_64-unknown-linux-gnu/stage0/bin/cargo /home/jyn/src/rust2/build 32
  Time (mean ± σ):     239.9 ms ±   8.0 ms    [User: 835.1 ms, System: 396.8 ms]
  Range (min … max):   231.1 ms … 257.1 ms    12 runs
```

on the slow machine:
```
Benchmark 1: build\\x86_64-pc-windows-msvc\\stage0-tools-bin\\rust-tidy.exe . build\\x86_64-pc-windows-msvc\\stage0/bin/cargo build 16
  Time (mean ± σ):      5.245 s ±  0.714 s    [User: 4.197 s, System: 7.256 s]
  Range (min … max):    4.156 s …  6.002 s    10 runs

Benchmark 1: build\\x86_64-pc-windows-msvc\\stage0-tools-bin\\rust-tidy.exe . build\\x86_64-pc-windows-msvc\\stage0/bin/cargo build 32
  Time (mean ± σ):      4.426 s ±  0.632 s    [User: 5.922 s, System: 9.719 s]
  Range (min … max):    2.920 s …  5.323 s    10 runs
```
So the difference between 2x and 8x is negligible on fast machines, and a significant improvement on slow machines. So I'd prefer to keep it.